File: java/L400/LeetCode0416.java
package L400;

public class LeetCode0416 {
    /**
     * 0-1 package   
     * 从数组中挑选一定的数，判断其和是否等于一半的总和
     */
    public boolean canPartitionInt(int[] nums) {
        int sum = 0;
        for (int n : nums){
            sum += n;
        }
        if (sum % 2 == 1){
            return false;
        }
        int half = sum >> 1;
        int[] arr = new int[half + 1];
        for (int i = 0;i < nums.length;i++){
            for (int j = half;j - nums[i] >= 0;j--){
                int curSum = nums[i] + arr[j - nums[i]];
                if (curSum == half){
                    return true;
                }
                if (curSum <= j){
                    arr[j] = curSum;
                }
            }
        }
        return false;
    }

    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int n : nums){
            sum += n;
        }
        if (sum % 2 == 1){
            return false;
        }
        int half = sum >> 1;
        boolean[] dp = new boolean[half + 1];
        dp[0] = true;
        for (int i = 0;i < nums.length;++i) {
            for (int j = half;j - nums[i] >= 0;--j) {
                dp[j] |= dp[j - nums[i]];
            }
        }
        return dp[half];
    }
}

File: cpp/L400/LeetCode0416.cc
#include <numeric>
#include <vector>
using std::vector;

class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum{std::accumulate(nums.begin(), nums.end(), 0)};
        if (sum % 2 != 0) {
            return false;
        }
        int half = sum / 2;
        vector<bool> dp(half + 1, false);
        dp[0] = true;
        for (int num : nums) {
            for (int i = half;i >= num;--i) {
                dp[i] = dp[i] || dp[i - num];
            }
        }
        return dp[half];
    }
};